common/src/main/java/org/conscrypt/OpenSSLAeadCipher.java

```diff
@@ -220,6 +220,52 @@ boolean allowsNonceReuse() {
         return false;
     }
 
+    /**
+     * Encrypts of decrypts data in a single-part operations or finishes multiple part operation.
+     * @param input Input data that is returned fully read
+     * @param output the storage buffer of the resulting encryption decryption. Output is returned position
+     *              incremented by bytesWritten Limit stays the same
+     * @return number of bytes the output buffer's position has moved by.
+     * @throws ShortBufferException if output.remaining() bytes are insufficient to hold the result.
+     * @throws IllegalBlockSizeException
+     * @throws BadPaddingException
+     */
+    @Override
+    protected int engineDoFinal(ByteBuffer input, ByteBuffer output) throws ShortBufferException, IllegalBlockSizeException, BadPaddingException {
+        if (input == null || output == null) {
+            throw new NullPointerException("Null ByteBuffer Error");
+        }
+        if ( getOutputSizeForFinal(input.remaining()) > output.remaining()) { // uses the child class's size if available
+            throw new ShortBufferException("Insufficient Bytes for Output Buffer");
+        }
+        if (bufCount != 0) {
+            return super.engineDoFinal(input, output);// traditional case
+        }
+        ByteBuffer tmp;
+        int bytesWritten;
+        if (!input.isDirect()) {
+            int incap = input.remaining(); // was capacity
+            tmp = ByteBuffer.allocateDirect(incap);
+            tmp.mark();
+            tmp.put(input);
+            tmp.reset();
+            input = tmp;
+        }
+        if (output.isDirect()) {
+            bytesWritten =  doFinalInternal(input, output);
+            output.position(output.position() + bytesWritten);
+            input.position(input.limit()); // API reasons
+        }
+        else {
+            tmp = ByteBuffer.allocateDirect(getOutputSizeForFinal(input.remaining()));
+            bytesWritten = doFinalInternal(input, tmp);
+            output.put(tmp);
+            input.position(input.limit()); // API reasons
+        }
+
+        return bytesWritten;
```

Comment:
It might be cleaner to rearrange this a bit.

- Move ByteBuffer tmp to be locals in the relevant if statements - inputClone for the copy of input, outputClone for the copy of output.

- Rearrange if statements to be if (!input.isDirect){...} if (!output.isDirect) {...} else {...} . It makes it a little clearer that your intention is to make Direct copies if needed, then do the doFinalInternal call.

Reply:
That makes more sense thank you!